### DISCUSSION TITLE
ATtiny1627 CuriosityNano - test of Arduino 6502 emulator + EhBasic - "implicit declaration of function" ERROR

### DISCUSSION THREAD

**Discussion by apws** [SpenceKonde/megaTinyCore#578] [Library discussion]
Hi, I today tested on few arduino boards this ino+c code: (zip download)
https://forum.arduino.cc/t/arduino-6502-emulator-basic-interpreter/188328
having fixed bug of nonexisting prog_uchar for "const unsigned char" as here:
https://forum.arduino.cc/t/gd2-h-329-27-error-prog_uchar-does-not-name-a-type/252021

It works OK on arduino leonardo and on sony spresense, but when I tried to compile this ino/c on tiny1627 CuriosityNano, this error was reported:
`W:\~prj\211223-arduino-6502emu\arduino_6502\cpu.c: In function 'read6502':
cpu.c:1060:15: error: implicit declaration of function 'getkey' [-Werror=implicit-function-declaration]
     tempval = getkey();
               ^~~~~~
cpu.c:1061:5: error: implicit declaration of function 'clearkey'; did you mean 'clearzero'? [-Werror=implicit-function-declaration]
     clearkey();
     ^~~~~~~~
     clearzero
`
can you please peek whats wrong where? tnx

**Comment by SpenceKonde** [comment]
It shouldn't work anywhere. cpu.c calls getkey **which is defined in the main sketch**.

It was skating by on account of Arduino's notoriously flaky function prototype generation would be my guess for why it didn't fail as it should have elsewhere. There should be a header with function declarations that is included by the ino and the c file, i'd say.

**Reply by apws** [reply]
ya, thanks, I am only scratching arduino now and was surprised where ino knows about c too ... okay ))

**Reply by apws** [reply]
uff, it is obvious, I am still totally beginner in C/C++ (just because of the complicated headers I am afraid of still, in fact for years) ... can you please point me, how to repair it with proper header? I tried something but with no success, in fact I first tried also to add the functions to "externs" but as there are both ino and c, I dont know exactly how to do proper COMMON header in this case... excuse me for asking for this, can you please help me? It seems that arduino original concept tries exactly that, to remove headers from simple things, encapsulating itself the ino subpart of entire C++ and providing some generation as you noted ... I understand that arduino is in fact C++/C, but now I am lost ... I know this is deeply cheeky question....

I pushed original code here:
https://github.com/apws/211225-arduino-6502-test

**Reply by apws** [reply]
I promise I will myself learn from your help to investigate more - but now, I am confused (which was probably initial idea of arduino to avoid )) ... back to learning C/C++, which would be good thing anyway... And I also consider buying some of your breakouts for new 1626 or 3226 when available, ordered some chips now for hacking already

**Reply by apws** [reply]
Or, if there exists some complete tutorial how to overcome such things in some incompatibilities of arduino sketeches compilable and noncompilabel elsewhere because of this missing includes, it would help me too, tnx

**Comment by SpenceKonde** [comment]
Sorry for the late reply , I have been in the midst of a very difficult move. I will try to take a look at this soon; I will have to fiddle with it. I can identify the general nature of the problem at a glance, but I can't identify exactly how to fix it without poking at it, and it's a fairly low priority since I have a critical bugfix release to do today for megaTinyCore (another one), need to finish unpacking my inventory to ship tindie orders with (thanks, by the way :-) ) but all that comes after making one of the last trips back to my old place to get the stuff that's still there before the landlord throws it away and bills us for disposal costs.

**Reply by apws** [reply]
many thanks, of course, I am also somewhat trying to solve it, but kicking would help... so, tnx :-)

**Reply by apws** [reply]
it seems I solved this only by adding this to cpu.c:
`extern uint8_t getkey();
extern void clearkey();
`
